TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/workflow/services/workflow_service.py

```diff
@@ -0,0 +1,860 @@
+                    depth=0
+                )
+
+        # Second pass: Establish parent-child relationships
+        for node_id, wf_node in workflow_nodes.items():
+            # Set parent based on boundaryID
+            if wf_node["boundary_id"] and wf_node["boundary_id"] in workflow_nodes:
+                wf_node["parent"] = wf_node["boundary_id"]
+
+        # Third pass: Calculate depth levels
+        def calculate_depth(node_id: str, visited: set) -> int:
+            """Recursively calculate depth for a node."""
+            if node_id in visited:
+                return 0  # Prevent infinite recursion
+            visited.add(node_id)
+
+            node = workflow_nodes.get(node_id)
+            if not node:
+                return 0
+
+            parent_id = node["parent"]
+            if not parent_id or parent_id not in workflow_nodes:
+                return 0
+
+            return 1 + calculate_depth(parent_id, visited)
+
+        for node_id in workflow_nodes:
+            workflow_nodes[node_id]["depth"] = calculate_depth(node_id, set())
+
+        # Return as list (will be sorted by _sort_nodes_intelligently)
+        return list(workflow_nodes.values())
+
+    def _sort_nodes_intelligently(self, nodes: List[WorkflowNode]) -> List[WorkflowNode]:
+        """Sort nodes considering both temporal and structural order.
+
+        Args:
+            nodes: List of WorkflowNode objects
+
+        Returns:
+            Sorted list of WorkflowNode objects
+        """
+        # Group nodes by depth level
+        nodes_by_depth: Dict[int, List[WorkflowNode]] = {}
+        for node in nodes:
+            depth = node["depth"]
+            if depth not in nodes_by_depth:
+                nodes_by_depth[depth] = []
+            nodes_by_depth[depth].append(node)
+
+        # Sort nodes at each depth level by start time
+        for depth in nodes_by_depth:
+            nodes_by_depth[depth].sort(
+                key=lambda x: (
+                    x.get("started_at") or "9999-99-99",  # Primary: start time
+                    x.get("finished_at") or "9999-99-99"  # Secondary: finish time
+                )
+            )
+
+        # Build final sorted list respecting depth hierarchy
+        sorted_nodes = []
+        for depth in sorted(nodes_by_depth.keys()):
+            sorted_nodes.extend(nodes_by_depth[depth])
+
+        return sorted_nodes
+
+    def _extract_workflow_steps(self, nodes: Dict[str, Any]) -> List[Dict[str, str]]:
+        """Extract step information from workflow nodes.
+
+        Args:
+            nodes: Dictionary of workflow nodes
+
+        Returns:
+            List of step dictionaries with name, phase, type, and started_at
+        """
+        steps = []
+        for node in nodes.values():
+            node_type = node.get("type", "")
+            # Include Pod and Suspend steps, plus Skipped steps to show conditional logic
+            if node_type in ["Pod", "Suspend", "Skipped"]:
```

Comment:
What do you want to filter out?  You'll filter out Resources, HttpRequests, Steps & Dags themselves.  What do you need to do this processing for?  For approve, you could just apply a filter on the whole set of nodes. I suspect that a lot of other usages might be like that too.  As it is now, some things might be skipped or otherwise never show up if they weren't skipped in a different run, which could be odd - making something not-skipped look like it was never even created at all.

I wouldn't call these steps - that's a very specific argo term.